ansible_collections/arista/avd/roles/eos_designs/python_modules/pool_manager/avdpoolmanager.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2023 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+from functools import lru_cache
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from yaml import SafeDumper, dump, safe_load
+
+from ansible_collections.arista.avd.plugins.filter.natural_sort import natural_sort
+from ansible_collections.arista.avd.plugins.plugin_utils.errors import AristaAvdError
+from ansible_collections.arista.avd.plugins.plugin_utils.utils import get, get_item
+
+if TYPE_CHECKING:
+    from ansible_collections.arista.avd.plugins.plugin_utils.eos_designs_shared_utils import SharedUtils
+
+FILE_HEADER = """\
+# This file is auto-generated by AVD eos_designs.
+# When there is a merge conflict, rebase the branch, accept all current changes, and re-run AVD.
+"""
+
+
+class AvdPoolManager:
+    """
+    Class used to handle pooled resources.
+
+    This class is imported and initialized once in eos_designs_facts
+    and given to shared_utils for each device.
+    """
+
+    def __init__(self, output_dir: str):
+        self.output_dir = Path(output_dir)
+        self.id_files: set[Path] = set()
+        self.changed_id_files: set[Path] = set()
+
+    def remove_stale_assignments(self):
+        """
+        Walk through all pools and in-place remove stale assignments and empty pools.
+
+        An assigment is deemed stale if "accessed" is not True.
+        The "accessed" field is also removed from the pool data.
+
+        Updates self.changed_id_files if any objects were removed from self.id_pool_data.
+
+        Note: This method is called from save_updated_pools and should _not_ be called manually.
+        """
+        for id_file in self.id_files:
+            id_data = self.id_pool_data(id_file)
+            for id_pool in id_data["id_pools"]:
+                # Remove id_assignments that were not accesses using "last_accessed".
+                id_assignments_count = len(id_pool["id_assignments"])
+                id_pool["id_assignments"] = [id_assignment for id_assignment in id_pool["id_assignments"] if id_assignment.pop("accessed", False)]
+                if id_assignments_count != len(id_pool["id_assignments"]):
+                    self.changed_id_files.add(id_file)
+
+            # Remove pools if there are no assignments left.
+            id_pools_count = len(id_data["id_pools"])
+            id_data["id_pools"] = [id_pool for id_pool in id_data["id_pools"] if id_pool["id_assignments"]]
+            if id_pools_count != len(id_data["id_pools"]):
+                self.changed_id_files.add(id_file)
+
+    def save_updated_pools(self, Dumper=SafeDumper):
```

Comment:
not sure having capitalized keywords argument is legit
```suggestion
    def save_updated_pools(self, dumper: <type>=SafeDumper) -> None:
```

Reply:
refactored everything so closing these old comments.